templates/contribute/index.md

```diff
@@ -3,53 +3,123 @@
 Here are some tips and tricks
 to make the process of contributing as smooth as possible.
 
-1. Use [Zulip](https://leanprover.zulipchat.com/) to
+* Use [Zulip](https://leanprover.zulipchat.com/) to
    discuss your contribution before and while you are working on it.
-2. Adhere to the guidelines:
+* Adhere to the guidelines:
    - The [style guide](style.html) for contributors.
    - The explanation of [naming conventions](naming.html).
    - The [documentation guidelines](doc.html).
-   
+
 Once you have code that you'd like to contribute, you should open a PR.
 There is a [video tutorial](https://www.youtube.com/watch?v=Bnc8w9lxe8A) walking you through the process of making a PR on YouTube.
-  
+
+## Working on mathlib
+
+We use `git` to manage and version control `mathlib`.
+The `master` branch is the "production" version of mathlib.
+It is essential that everything in the master branch compiles without errors, and there are no `sorry`s.
+To ensure this, we only commit changes to `master` that have passed CI tests, and have been approved by mathlib maintainers.
+
+While you're working on a new contribution to `mathlib`, you should do this on a different branch.
+It's okay to do this in your own fork of the `mathlib` repository,
+or you can introduce yourself on Zulip and ask for write access to non-`master` branches of the mathlib repository.
+Please include your GitHub username in your request.
+It's polite to prefix the branch name with your username, so it's easier for us to clean up clutter.
+(Once you're making a pull request, we'll ask you to do so from a branch of the mathlib repository,
+rather than from your own fork, as CI works better this way.)
+
+Typical workflow:
+* To get started, you'll need a local copy of mathlib.
+* At https://github.com/leanprover-community/mathlib, click "Fork" in the top right, to make your own fork of the repository.
+  Your fork is at https://github.com/USER/mathlib.
+  Alternatively, if you've asked for write access you can just use https://github.com/leanprover-community/mathlib.
+* Now make a local clone of the repository.
+  ```
+  git clone https://github.com/USER/mathlib.git
+  cd mathlib
+  ```
+* The steps above only need to be done once (not once for each contribution).
+* Now, each time you want to work on a new change to mathlib, create a new branch:
+  ```
+  git checkout -b my_new_branch   # This creates a new branch
```

Comment:
```suggestion
  git checkout -b my_new_branch master  # This creates a new branch departing from the master branch
```
Explicitly adding `master` might avoid confusion when someone follows these steps for a 2nd contribution, if they still have an old branch checked out. Of course, there is a risk that `master` is very old. So maybe we should talk about `git fetch` as well...

Reply:
I haven't read all the new file but that part looks like it should talk about `leanproject pr my_new_branch`.

Reply:
This PR has been sitting for quite a while, and looks good to me except for this comment. I don't use `leanproject` for this so I don't know for sure, is `leanproject pr my_new_branch` the same as `git checkout -b my_new_branch && git reset --hard origin/master && leanproject get-cache`?

Reply:
What is does is equivalent to what you wrote + setting upstream tracking branch (so that `git push` doesn't complain).

Reply:
```suggestion
  leanproject pr my_new_branch
```

Reply:
Note that `leanproject pr` also moves the local `master` reference, which is likely to be surprising if someone has made some junk commits to their local master.

Reply:
Hmm, I would prefer if this instruction was as unsurprising as possible

Reply:
I don't understand why Eric is complaining here. If people mess up with their master branch then they will pay for it anyway. This is orthogonal to the discussion.